### PR TITLE
[ci, docs, model] feat: add layered test architecture (L0-L2) for new model onboarding

### PR DESCRIPTION
**Pull request by TimYangst**
### What does this PR do?

Introduces a structured, layered test architecture (L0–L2) to streamline new model onboarding and improve test coverage across the VeOmni codebase.

**Core additions:**

- Add `veomni/testing/` package with reusable utilities (`TensorComparator`, dummy data generators, toy config helpers) importable by downstream projects
- Add `tests/conftest.py` with pytest markers (`L0`, `L1`, `L2`, `v4_only`, `v5_only`, `multi_gpu`, `moe`) for fine-grained test selection

**New tests by level:**

- **L0 (unit, CPU/single-GPU):** HF processor alignment validation, fused cross-entropy kernel correctness, patchgen equivalence, padding edge cases
- **L1 (distributed correctness, multi-GPU):** single-GPU vs FSDP2 equivalence, parallelism combination consistency (FSDP+SP+EP), dummy forward validation, tied embeddings correctness
- **L2 (end-to-end):** toy-config training pipeline (no model download required)

**Test restructuring:**

- Consolidate distributed tests into `tests/distributed/` (moved checkpoint, parallel tests from `tests/checkpoints/` and `tests/e2e/`)
- Rename utility modules for clarity (`utils.py` → `data_test_fixtures.py`, `e2e_fixtures.py`, `model_patch_helpers.py`, `checkpoint_helpers.py`, `distributed_test_helpers.py`)
- Remove obsolete `tests/tools/common_utils.py` and `tests/checkpoints/utils.py`
- Add level markers to all existing test files

**CI & docs:**

- Update GPU/NPU CI workflows to run new L1 distributed alignment tests and L2 toy e2e tests under both transformers v4 and v5
- Update docs: v5 onboarding checklist and testing guide with L0-L2 hierarchy

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] PR title follows `[{modules}] {type}: {description}` format

### Test

All new tests are integrated into CI workflows (`gpu_e2e_test.yml`, `gpu_unit_tests.yml`, `npu_e2e_test.yml`, `npu_unit_tests.yml`) and run under both transformers v4 (stable) and v5 (experimental).

### API and Usage Example

```python
# Import reusable test utilities
from veomni.testing import TensorComparator, get_dummy_data, get_toy_config_path

# Compare tensors with tolerance
comparator = TensorComparator()
comparator.assert_close(tensor_a, tensor_b, atol=1e-5, rtol=1e-5)

# Generate dummy training data
data = get_dummy_data(batch_size=4, seq_len=128)

# Get toy config path for e2e tests (no model download)
config_path = get_toy_config_path("qwen3_moe")
```

### Design & Code Changes

- **`veomni/testing/`**: New package with `comparison_utils.py` (TensorComparator, assert helpers), `data_generators.py` (dummy data factory), `toy_config_utils.py` (toy config loader/validator)
- **`tests/conftest.py`**: Central pytest marker registration and transformers version skip helpers
- **`tests/distributed/`**: New directory consolidating all distributed/parallel tests with shared helpers (`distributed_test_helpers.py`, `checkpoint_helpers.py`)
- **`tests/e2e/e2e_fixtures.py`**: Refactored from `utils.py` with improved fixture organization
- **CI workflows**: Added steps for L1 distributed alignment tests and L2 toy e2e tests
- **`pyproject.toml`**: Added test-related dependencies

### Checklist Before Submitting

- [ ] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- [ ] Applied pre-commit checks
- [ ] Added/updated documentation
- [ ] Added tests to CI workflow (or explained why not feasible)

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    config.addinivalue_line("markers", "L0: Level 0 - Data processing & transform correctness (no GPU)")
+    config.addinivalue_line("markers", "L1: Level 1 - Kernel & operator consistency (single GPU)")
+    config.addinivalue_line("markers", "L2: Level 2 - Single-GPU vs FSDP equivalence (2+ GPUs)")
+    config.addinivalue_line("markers", "L3: Level 3 - Parallelism combination consistency (2-8 GPUs)")
+    config.addinivalue_line("markers", "L4: Level 4 - Corner cases & robustness")
+    config.addinivalue_line("markers", "L5: Level 5 - Integration / smoke test")
+    config.addinivalue_line("markers", "v4_only: Requires transformers < 5.0.0")
+    config.addinivalue_line("markers", "v5_only: Requires transformers >= 5.0.0")
+    config.addinivalue_line("markers", "multi_gpu: Requires multiple GPUs")
+    config.addinivalue_line("markers", "moe: MoE model specific test")
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply tests/conftest.py]: Updated

File: tests/distributed/distributed_test_helpers.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+class ParallelConfig:
+    """Describes a parallelism configuration for distributed tests."""
+
+    sp_size: int = 1
+    ep_size: int = 1
+    fsdp_mode: str = "fsdp2"
+
+    @property
+    def world_size(self) -> int:
+        return max(self.sp_size * 2, 2)
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply tests/distributed/distributed_test_helpers.py]: Updated

File: tests/models/test_patchgen_equivalence.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    from veomni.models import build_foundation_model
+
+    set_seed(42)
+    device = get_device_type()
+
+    # Build HF baseline model
+    os.environ["MODELING_BACKEND"] = "hf"
+    from veomni import _apply_patches
+
+    _apply_patches()
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply tests/models/test_patchgen_equivalence.py:97]: _apply_patches in test_patchgen_equivalence.py - The design is inherent to what the test does (compare two backends). Using monkeypatch would be slightly cleaner but wouldn't change the fundamental approach. Low priority.

File: tests/e2e/e2e_fixtures.py [hunk 4]
```diff
@@ -184,7 +215,10 @@
     command_list = []
     for task in test_tasks:
         for mode in model_modes:
-            port = 12345 + random.randint(0, 100)
+            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+                s.bind(("", 0))
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply tests/e2e/e2e_fixtures.py]: Socket bind to all interfaces - Standard test pattern, not a real security concern. Low priority.
